### PR TITLE
cppad: add v20180000.0, v20240000.4, master

### PR DESCRIPTION
**Pull request by whart222**
Several changes
1. Updating to latest version
2. Disabling documentation generation (which causes build errors)

### CODE REVIEW

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py]: I get the following with the new URL:

```suggestion
    version(
        "20180000.0", sha256="a5226e4c5aa2ad6d95668f987b39939315bf134a0a793231984e6d42d6488cca"
    )
```

Any idea why that might be the case?

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py]: I cannot confirm the sha256 for this version, not with the new URL or the old one.  Looks like it isn't in the GitHub repo.

Perhaps it should be deprecated?  See https://spack.readthedocs.io/en/latest/packaging_guide.html#deprecating-old-versions.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py]: The name of the version should match that of the branch (https://spack.readthedocs.io/en/latest/packaging_guide.html#git).  For one thing, it is more intuitive for people who look at the repository before building with Spack.

For more on how Spack compares versions, see https://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py]: That URL contains a link to a page that results in `Not Found`.  Perhaps it would be better to link to the readthdocs page at `https://cppad.readthedocs.io/latest/`?

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Confirmed the two new version sha256s but have issues with the older ones and the homepage.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: I didn't edit this sha value.  Looking at Github, it appears that the tarball was copied from a previous site.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: I deprecated both of the two old versions. Do you think I should update their sha256 values based on the current github tarballs? I don't know who setup these versions initially.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: I made this change in the commit.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: I updated the URL to point to the Github page for cppad.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (alecbcs) [inline var/spack/repos/builtin/packages/cppad/package.py]:
```suggestion
    homepage = "https://github.com/coin-or/CppAD"
    url = "https://github.com/coin-or/CppAD/archive/refs/tags/20240000.4.tar.gz"
    git = "https://github.com/coin-or/CppAD.git"
    
    maintainers("whart222")

```

Following the common format, the `maintainers()` directive should go under homepage, url, and git variables and before the first `version()` directive.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (alecbcs) [inline var/spack/repos/builtin/packages/cppad/package.py]: I don't fully understand the changes to the cmake args. Does the package no longer support installing the docs or was this comment left in from debugging? Additionally I'd recommend replacing,

```
args.append("-DCMAKE_BUILD_TYPE=Release")
```
with
```
self.define("CMAKE_BUILD_TYPE", "Release")
```
inside of the args list declaration.

**Review by alecbcs — CHANGES_REQUESTED** [review]

**Comment by alecbcs** [comment]
@whart222 this PR looks great! Just a couple additional recommendations :)

@tldahlgren feel free to continue as lead. Just saw a couple additional points of feedback I didn't want missed.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/cppad/package.py]: Alec is correct in that the convention is to have directives appear **after** properties.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py]: I still debate whether this should be changed to the new sha256 despite the deprecation:

```suggestion
        sha256="a5226e4c5aa2ad6d95668f987b39939315bf134a0a793231984e6d42d6488cca",
```

**Review by tldahlgren — COMMENTED** [review]
Changes look good.  Just a couple of questions/points.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: I'm good with this change.  I just pushed this fix.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: Done.

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (whart222) [reply var/spack/repos/builtin/packages/cppad/package.py]: When I try to install cppad locally, it fails to install with docs.  IMHO, this isn't an essential feature of the installation since the docs are available online.  Hence, I'm inclined to omit them.

Perhaps building with docs requires other packages than what I've installed?  If so, then we should conditionally build docs after checking for those.

**Comment by whart222** [comment]
I think I'm done with revisions here.  Is there something I missed?

**Review by tldahlgren — APPROVED** [review]

File: var/spack/repos/builtin/packages/cppad/package.py
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/cppad/package.py:44]: We usually suggest removing comments like this UNLESS there is a reason for keeping it, which I will assume that's the intention behind the comment.

**Comment by tldahlgren** [comment]
@alecbcs It looks like you need to be involved here since you have outstanding changes that are blocking merging.

**Review by alecbcs — APPROVED** [review]
Looks good to me. Apologies for the delay getting back to this.

**Pull request merged**
